File: apps/bike-map/src/app/core/services/geocoding.service.ts
import { Injectable, inject } from '@angular/core';
import { GeoPoint } from '../models/geo.types';
import { OfflineNetworkService } from './offline-network.service';
import { StorageService } from './storage.service';

export interface SearchResultItem {
  id: string;
  name: string;
  subText: string;
  category: 'address' | 'landmark' | 'poi' | 'city' | 'frequent';
  point: GeoPoint;
  houseNumber?: string;
  streetName?: string;
}

@Injectable({
  providedIn: 'root'
})
export class GeocodingService {
  private networkService = inject(OfflineNetworkService);
  private storageService = inject(StorageService);

  /**
   * Returns top frequent and recent locations for instant suggestion when focusing on search
   */
  getFrequentAndRecentPlaces(): SearchResultItem[] {
    const frequent = this.storageService.getFrequentPlaces();
    return frequent.map(fp => ({
      id: `fp_${fp.id}`,
      name: fp.name,
      subText: fp.subText || 'Saved frequent location',
      category: 'frequent',
      point: fp.point
    }));
  }

  /**
   * Reverse geocodes coordinates into "Street Name, Number" (e.g. "Rua Silvino Lopes, 240")
   */
  async reverseGeocode(lat: number, lng: number): Promise<{ name: string; street?: string; houseNumber?: string }> {
    try {
      const url = `https://nominatim.openstreetmap.org/reverse?format=json&addressdetails=1&lat=${lat}&lon=${lng}&zoom=18`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 3500);
      const res = await fetch(url, {
        signal: controller.signal,
        headers: { 'Accept-Language': 'pt-BR,pt;q=0.9,en;q=0.8' }
      });
      clearTimeout(timeoutId);

      if (res.ok) {
        const data = await res.json();
        if (data && data.address) {
          const addr = data.address;
          const road = addr.road || addr.pedestrian || addr.street || addr.avenue || addr.square || addr.neighbourhood;
          const houseNum = addr.house_number;
          const sub = addr.suburb || addr.neighbourhood || addr.city_district || '';

          if (road) {
            const formatted = houseNum ? `${road}, ${houseNum}` : road;
            const full = sub ? `${formatted}, ${sub}` : formatted;
            return { name: full, street: road, houseNumber: houseNum };
          } else if (data.name) {
            return { name: data.name };
          } else if (data.display_name) {
            return { name: data.display_name.split(',')[0] };
          }
        }
      }
    } catch {}

    // Fallback: Photon reverse
    try {
      const pUrl = `https://photon.komoot.io/reverse?lat=${lat}&lon=${lng}`;
      const pRes = await fetch(pUrl);
      if (pRes.ok) {
        const pData = await pRes.json();
        if (pData?.features?.[0]?.properties) {
          const props = pData.features[0].properties;
          const road = props.street || props.name;
          const houseNum = props.housenumber;
          if (road) {
            return { name: houseNum ? `${road}, ${houseNum}` : road, street: road, houseNumber: houseNum };
          }
        }
      }
    } catch {}

    return { name: `Location (${lat.toFixed(4)}, ${lng.toFixed(4)})` };
  }

  /**
   * Searches places, streets, addresses with house numbers, and POIs globally
   */
  async searchPlaces(query: string, userLat?: number, userLng?: number): Promise<SearchResultItem[]> {
    const q = query.trim();
    if (!q) {
      return this.getFrequentAndRecentPlaces();
    }

    const results: SearchResultItem[] = [];

    // 0. Include matching frequent places first
    const frequent = this.storageService.getFrequentPlaces();
    for (const fp of frequent) {
      if (fp.name.toLowerCase().includes(q.toLowerCase()) || fp.subText.toLowerCase().includes(q.toLowerCase())) {
        results.push({
          id: `fp_${fp.id}`,
          name: fp.name,
          subText: `⭐ Frequent: ${fp.subText}`,
          category: 'frequent',
          point: fp.point
        });
      }
    }

    // 1. Search local POIs
    const localPois = this.networkService.pois();
    for (const poi of localPois) {
      if (poi.name.toLowerCase().includes(q.toLowerCase()) || poi.category.includes(q.toLowerCase())) {
        results.push({
          id: `poi_${poi.id}`,
          name: poi.name,
          subText: poi.description || `Category: ${poi.category}`,
          category: 'poi',
          point: poi.coordinates
        });
      }
    }

    // Extract house number if user entered one (e.g. ", 123", " nº 450", " 100A")
    const extractedNumMatch = q.match(/(?:n[º°o]?\s*|\,\s*|\s+)(\d+[a-zA-Z]?)(?:\s|$)/i);
    const queryHouseNum = extractedNumMatch ? extractedNumMatch[1] : null;

    // 2. Global OpenStreetMap Photon Search
    try {
      let url = `https://photon.komoot.io/api/?q=${encodeURIComponent(q)}&limit=8`;
      if (userLat !== undefined && userLng !== undefined) {
        url += `&lat=${userLat}&lon=${userLng}`;
      }

      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 3500);
      const res = await fetch(url, { signal: controller.signal });
      clearTimeout(timeoutId);

      if (res.ok) {
        const data = await res.json();
        if (data && data.features) {
          for (const feat of data.features) {
            const props = feat.properties || {};
            const coords = feat.geometry?.coordinates; // [lng, lat]
            if (!coords || coords.length < 2) continue;

            const houseNum = props.housenumber || queryHouseNum;
            let formattedTitle = '';

            if (props.street) {
              formattedTitle = houseNum ? `${props.street}, ${houseNum}` : props.street;
            } else if (props.name) {
              formattedTitle = houseNum && !props.name.includes(houseNum) ? `${props.name}, ${houseNum}` : props.name;
            } else if (props.city) {
              formattedTitle = props.city;
            } else {
              formattedTitle = q;
            }

            const subParts = [
              props.district || props.suburb,
              props.city || props.town || props.village,
              props.state,
              props.postcode,
              props.country
            ].filter(Boolean);

            const subText = subParts.join(', ') || 'Global Location';

            results.push({
              id: `osm_${coords[1].toFixed(5)}_${coords[0].toFixed(5)}_${houseNum || ''}`,
              name: formattedTitle,
              subText,
              category: props.osm_value === 'city' ? 'city' : (props.street || houseNum) ? 'address' : 'landmark',
              houseNumber: houseNum || undefined,
              streetName: props.street || undefined,
              point: {
                lat: Number(coords[1].toFixed(6)),
                lng: Number(coords[0].toFixed(6)),
                ele: 20
              }
            });
          }
        }
      }
    } catch {
      // 3. Fallback: OpenStreetMap Nominatim with structured addressdetails
      try {
        let nomUrl = `https://nominatim.openstreetmap.org/search?format=json&addressdetails=1&q=${encodeURIComponent(q)}&limit=5`;
        if (userLat !== undefined && userLng !== undefined) {
          nomUrl += `&viewbox=${userLng - 0.5},${userLat + 0.5},${userLng + 0.5},${userLat - 0.5}&bounded=0`;
        }

        const nomRes = await fetch(nomUrl);
        if (nomRes.ok) {
          const nomData = await nomRes.json();
          for (const item of nomData) {
            const addr = item.address || {};
            const houseNum = addr.house_number || queryHouseNum;
            const road = addr.road || addr.pedestrian || addr.street;

            let title = item.name || '';
            if (road) {
              title = houseNum ? `${road}, ${houseNum}` : road;
            } else if (!title) {
              title = item.display_name.split(',')[0];
            }

            results.push({
              id: `nom_${item.place_id}`,
              name: title,
              subText: item.display_name,
              category: 'address',
              houseNumber: houseNum,
              streetName: road,
              point: {
                lat: parseFloat(item.lat),
                lng: parseFloat(item.lon),
                ele: 20
              }
            });
          }
        }
      } catch {}
    }

    return results;
  }
}

File: app/src/main/kotlin/com/ebike/router/service/GeocodingService.kt
package com.ebike.router.service

import com.ebike.router.model.GeoPoint
import com.ebike.router.model.PlaceCategory
import com.ebike.router.model.SearchResultItem
import com.google.gson.Gson
import com.google.gson.JsonObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

class GeocodingService {
    private val client = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(5, TimeUnit.SECONDS)
        .build()

    private val gson = Gson()

    suspend fun searchPlaces(
        query: String,
        userLat: Double? = null,
        userLng: Double? = null
    ): List<SearchResultItem> = withContext(Dispatchers.IO) {
        val q = query.trim()
        if (q.length < 2) return@withContext emptyList()

        val results = mutableListOf<SearchResultItem>()

        // 1. Photon (Komoot OSM)
        try {
            var url = "https://photon.komoot.io/api/?q=${URLEncoder.encode(q, "UTF-8")}&limit=6"
            if (userLat != null && userLng != null) {
                url += "&lat=$userLat&lon=$userLng"
            }

            val request = Request.Builder()
                .url(url)
                .header("User-Agent", "EBikeRouterAndroid/1.0")
                .build()

            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                val body = response.body?.string()
                if (body != null) {
                    val root = gson.fromJson(body, JsonObject::class.java)
                    val features = root.getAsJsonArray("features")
                    if (features != null) {
                        for (i in 0 until features.size()) {
                            val feat = features.get(i).asJsonObject
                            val props = feat.getAsJsonObject("properties")
                            val geom = feat.getAsJsonObject("geometry")
                            val coords = geom?.getAsJsonArray("coordinates")

                            if (coords != null && coords.size() >= 2) {
                                val lng = coords.get(0).asDouble
                                val lat = coords.get(1).asDouble

                                val name = props?.get("name")?.asString
                                    ?: props?.get("street")?.asString
                                    ?: props?.get("city")?.asString
                                    ?: q

                                val parts = listOfNotNull(
                                    props?.get("street")?.asString,
                                    props?.get("district")?.asString,
                                    props?.get("city")?.asString,
                                    props?.get("state")?.asString,
                                    props?.get("country")?.asString
                                )
                                val subText = if (parts.isNotEmpty()) parts.joinToString(", ") else "Localização Global"

                                val osmVal = props?.get("osm_value")?.asString
                                val category = when {
                                    osmVal == "city" -> PlaceCategory.CITY
                                    props?.has("street") == true -> PlaceCategory.ADDRESS
                                    else -> PlaceCategory.LANDMARK
                                }

                                results.add(
                                    SearchResultItem(
                                        id = "osm_${lat}_${lng}_$i",
                                        name = name,
                                        subText = subText,
                                        category = category,
                                        point = GeoPoint(lat, lng, 20.0)
                                    )
                                )
                            }
                        }
                    }
                }
            }
        } catch (_: Exception) {}

        if (results.isNotEmpty()) return@withContext results

        // 2. Nominatim fallback
        try {
            val nomUrl = "https://nominatim.openstreetmap.org/search?format=json&q=${URLEncoder.encode(q, "UTF-8")}&limit=5"
            val request = Request.Builder()
                .url(nomUrl)
                .header("User-Agent", "EBikeRouterAndroid/1.0")
                .build()

            val response = client.newCall(request).execute()
            if (response.isSuccessful) {
                val body = response.body?.string()
                if (body != null) {
                    val arr = gson.fromJson(body, com.google.gson.JsonArray::class.java)
                    for (i in 0 until arr.size()) {
                        val item = arr.get(i).asJsonObject
                        val lat = item.get("lat").asString.toDouble()
                        val lon = item.get("lon").asString.toDouble()
                        val displayName = item.get("display_name").asString
                        val title = displayName.split(",").firstOrNull() ?: q

                        results.add(
                            SearchResultItem(
                                id = "nom_${item.get("place_id")?.asString ?: i}",
                                name = title.trim(),
                                subText = displayName.trim(),
                                category = PlaceCategory.ADDRESS,
                                point = GeoPoint(lat, lon, 20.0)
                            )
                        )
                    }
                }
            }
        } catch (_: Exception) {}

        return@withContext results
    }
}
